### ISSUE TITLE
Tags, categories, post formats

### ISSUE THREAD

**Issue by marcelomazza** [MichaelArestad/Press-This#45]
The mockup says it all!

![press-this_post-format](https://cloud.githubusercontent.com/assets/384622/6098818/fdc2bc62-afe9-11e4-867e-11eb73028a82.jpg)

Let me know what you think.

**Comment by MichaelArestad** [comment]
A. Still want content blocks to eventually be built in so props for the reference.

B. This could work, but it would be hard to change later. Also, would the interface totally change per type? Could a user still add/edit text to go with the image/gallery, whatever?

**Comment by MichaelArestad** [comment]
I finally have some mockups for adding and changing tags, categories, post formats on mobile.

### Getting there
Tapping the tag icon on the top right would open the panel for the meta stuff

![image](https://cloud.githubusercontent.com/assets/1123119/6102156/fa8175f8-afed-11e4-99e1-be1e46e9fd9b.png)


### Post formats, categories, tags
Panel with various settings
![image](https://cloud.githubusercontent.com/assets/1123119/6102173/20e0ddd8-afee-11e4-8c64-5a53ce95aafc.png)

I could have squeezed all of these settings into this one screen, but dropdowns and autocompletes start getting messy to use all on one screen. The added benefit of each item getting its own screen is extra clarity and space.

### Upon tapping post formats
![image](https://cloud.githubusercontent.com/assets/1123119/6102182/37387082-afee-11e4-86e0-3d818221bced.png)


### The categories and tags screen would be the same (with autocomplete)
![image](https://cloud.githubusercontent.com/assets/1123119/6102189/4bce1bfa-afee-11e4-8d83-f793440ca4b0.png)

### On desktop
Still mocking up, but would be pretty similar. The major change is the post formats, categories, tags would all be visible in a sidebar from the start.

### Thoughts?
CC @stephdau @marcelomazza @melchoyce

**Comment by stephdau** [comment]
Not sure if categories will make it in for Wednesday, but we can take a stab at the rest. It mostly depends on what else is left.

@azaozz : if we go that route, could we split the job as:
 * I'd handle the main panel and post formats,
 * You'd handle the tags & its autocompletion, since you mentioned you had ideas how to?

**Comment by danielbachhuber** [comment]
Dig it — similar to Poster's metadata management UI

**Comment by marcelomazza** [comment]
> B. This could work, but it would be hard to change later. Also, would the interface totally change per type? Could a user still add/edit text to go with the image/gallery, whatever?

@MichaelArestad I wasn't thinking in changing the interface per type, but more in giving different entry points to the same interface, without change it. For example, if you tap on image, you get the Media Uploader, but after you load the image you would end up in the Post Editor, as we have it now, with the picture loaded.

And my feedback for your mocks! (I love them, congrats)

##### About tags, categories, post formats
I really like how you resolved adding features without adding complexity to the initial UI. I would love to see how are you envisioning this on desktop without cluttering the layout.

##### iOS language
I don't know if we care but the "Done" at the top right and the caret (">") in each item have a really iOS feel, maybe we should look for a more platform agnostic language. However I feel they work well. Same for the back button on top right.

##### Multiple screens instead of one
I like that approach, I think we need to be conscious about the browser state, so if the user hit the "back" of the browser can navigate each screen instead of going to a different URL.

On that line, have you thought on "accordion" instead of "drill-down" pages?

##### Categories autocomplete
I think @azaozz mentioned something about the categories didn't have autocomplete and it was just a list, but I'm not sure. He'll clarify for sure.

**Comment by melchoyce** [comment]
I'm a little lost — did you get to the list screen by tapping the tag icon?

The list screen and tag/cat screens seem on point. What's the touch area for removing an icon? Tapping just the icon, or also the label?

**Comment by MichaelArestad** [comment]
> did you get to the list screen by tapping the tag icon?

Yep. Though it contains more than just tags, it seemed to convey more about what it contains than a settings icon.

> What's the touch area for removing an icon?

For removing a tag/category, the touch area could be the whole word or I'd make the delete icon a 40 x 40 (invisible) tap target.

**Comment by MichaelArestad** [comment]
@melchoyce quick video: https://cloudup.com/c9LwWJ6FYDE

**Comment by MichaelArestad** [comment]
Categories design update:

![image](https://cloud.githubusercontent.com/assets/1123119/6139743/7e543208-b14c-11e4-9fa1-ca245870786a.png)

**Comment by marcelomazza** [comment]
I like the addition of the magnifier on the top right! In this case, I would remove the cross in "new category" to reinforce the idea of toggle views (the views should be switched from the same place only).

**Comment by MichaelArestad** [comment]
Updated Categories:

![image](https://cloud.githubusercontent.com/assets/1123119/6154204/18848cd4-b1e7-11e4-90af-a2e6eacad027.png)

![image](https://cloud.githubusercontent.com/assets/1123119/6154214/26c9ce08-b1e7-11e4-9a84-80fa1d36278c.png)

**Comment by marcelomazza** [comment]
I love this final iteration, I wouldn't change anything. Just as a reference, you can check the thought process we did between @MichaelArestad @folletto and I here:

https://wordpress.slack.com/archives/feature-pressthis/p1423610523003051

These mockups are excellent! :+1:

**Comment by marcelomazza** [comment]
@stephdau we all agreed on the [mockups](https://github.com/MichaelArestad/Press-This/issues/45#issuecomment-73943955) above.

Chatted in Slack about this with drew, azaozz, stephdau and michaelarestad.

The only thing to be added would be indentation to represent the hierarchy.

--
[Full discussion in Slack for reference](https://wordpress.slack.com/archives/feature-pressthis/p1423772419003844)

**Comment by MichaelArestad** [comment]
Indentation is so dang doable. How many levels should I prepare to indent? 5?

**Comment by MichaelArestad** [comment]
Maybe something like this:

![image](https://cloud.githubusercontent.com/assets/1123119/6181457/3b29b250-b2f0-11e4-9db5-d85a1cde44f2.png)

**Comment by folletto** [comment]
Superb work :+1:

**Comment by stephdau** [comment]
Started implementation in 440ffdf0a4, with functional save.

**Comment by MichaelArestad** [comment]
Category search ideas.

I still think the best way to show the results with the hierarchy is to include the parents. I'm not sure how performant it would be.

A. greyed out compact parents (highlights results better
![image](https://cloud.githubusercontent.com/assets/1123119/6192205/e5789b00-b36c-11e4-8907-b282584e9032.png)

B. No style changes
![image](https://cloud.githubusercontent.com/assets/1123119/6192211/ee18db3a-b36c-11e4-9382-a93730594ee6.png)

**Comment by folletto** [comment]
I'd do the grayed out if not technically too complex, I'd just maybe suggest a slightly different gray?

![screen shot 2015-02-13 at 17 49 08](https://cloud.githubusercontent.com/assets/4389/6192330/a43734d0-b3a8-11e4-945f-de5f4d5abc0a.png)

**Comment by melchoyce** [comment]
When I look at the greyed out mockups, I honestly feel like I have no idea what's going on. The screen is starting to get pretty complex and confusing to me. I even think the grey indented child categories are a little hard to follow, since they almost seem even more separate from their parents. 

What do you think of something more like this?
![categories](https://cloud.githubusercontent.com/assets/2846578/6192481/87d89126-b380-11e4-985c-53f14bad9b4d.jpg)

Random idea: what if we combined the search and add category field into one? If you search and it finds the category, it selects it. Otherwise, it adds your search term as a new category.

**Comment by MichaelArestad** [comment]
@folletto 

> I'd do the grayed out if not technically too complex, I'd just maybe suggest a slightly different gray (i.e. blended in the background, since they are not selectable in any way and just context)?

Maybe. It makes it a bit harder to follow the indent trail that way.

@melchoyce 

> What do you think of something more like this?

I was considering something like this, but it requires quite a bit of horizontal space.

Side note: That arrow has always thrown me off. It just means "reply" to my brain.
 
> Random idea: what if we combined the search and add category field into one? If you search and it finds the category, it selects it. Otherwise, it adds your search term as a new category.

I like this idea, but gotta make sure it makes sense to users. Chewing on this while I go get a burrito.

**Comment by folletto** [comment]
True, but I think it's not super relevant in that context, since you just need it to read the context. However you might do the same and keeping a subtle outline to show the indent level. :)

**Comment by melchoyce** [comment]
> I was considering something like this, but it requires quite a bit of horizontal space.

I don't think the majority of users really go beyond two or three levels of indentation, so I'm not really worried about horizontal space. But for sure, we could still tighten it up a bit.

> Side note: That arrow has always thrown me off. It just means "reply" to my brain.

That's fair. We could definitely explore a different style of indicating the relationship.

**Comment by marcelomazza** [comment]
I'd say that we can play it simple here:

### Just indentation for the hierarchy:
![categories_1a](https://cloud.githubusercontent.com/assets/384622/6193522/6f57fa60-b3bc-11e4-90e0-af0d71fd41aa.jpg)

### Leave the indentation for the parents, and show them with less opacity:
![categories_1b](https://cloud.githubusercontent.com/assets/384622/6193560/bee085f2-b3bc-11e4-83c2-f4472420ff4f.jpg)

Someone can say that less opacity gives a "disabled" feel, but I think that's not a problem because: a) if the user is typing "Cat" why would she/he want to tap on "Design"? b) Even if they tap in a parent we'd leave the functionality and you'll be able to toggle parents there too.

**Comment by stephdau** [comment]
@mel: "the majority of users" isn't who complains about those things. ;)

**Comment by melchoyce** [comment]
> "the majority of users" isn't who complains about those things. ;)

So we should start designing for the vocal minority?

**Comment by melchoyce** [comment]
For child categories: what about display them as:

`Cat gifs (in "Cats")`

**Comment by folletto** [comment]
> I'd say that we can play it simple here

I'd go for this only if it simplifies drastically the code to be done, otherwise this isn't very effective because the height isn't reduced and it's not highlighting the things you can actually tap. Ideally search should bring the item you want straight under the search field, so I feel the "disabled" items need to fade away entirely (thus the concept I expressed above of using the background, but can be anything that reduces height and removes the perception of being buttons). :)

> For child categories: what about display them as `Cat gifs (in "Cats")`

I think this is another very viable approach that would remove entirely the height issue I just mentioned above, using maybe a secondary stile for the `(in "Cats")` part. :)

**Comment by MichaelArestad** [comment]
> I think this is another very viable approach that would remove entirely the height issue I just mentioned above, using maybe a secondary stile for the (in "Cats") part. :)

That kind of thing occurred to me, but it would end up looking like this:

`[Font hacks in Web Typography in Web design in Design]`
or
`[Font hacks in Web Typography > Web design > Design]` (confusing)

both of which would get truncated or have to go multiline (shudder)

Another thought is a single row above with the hierarchy chain.
```
[Design > Web design > Web Typography]
[Font hacks]
```

but the clearest and simplest solution may just be:
![image](https://cloud.githubusercontent.com/assets/1123119/6194889/f1c95c00-b385-11e4-8e2b-fa92e64887c0.png)


And this way, the user can still add/remove parent categories.

**Comment by MichaelArestad** [comment]
Another spinoff with clear "selected" states:

![image](https://cloud.githubusercontent.com/assets/1123119/6194969/c6738688-b386-11e4-8ad8-870d2603ed37.png)

**Comment by MichaelArestad** [comment]
@marcelomazza 

> Someone can say that less opacity gives a "disabled" feel, but I think that's not a problem because: a) if the user is typing "Cat" why would she/he want to tap on "Design"? b) Even if they tap in a parent we'd leave the functionality and you'll be able to toggle parents there too.

I would want to tap every parent category to go with the category I searched for. The simple version (without horizontal lines) is my favorite so far.

**Comment by marcelomazza** [comment]
@MichaelArestad yeap, I understand the feeling. The reasoning behind the low opacity is to not grab the attention of the eye, and make a scan line through the matched categories. But it can definitely can have side effects.

I like your last two mockups, I would go with them (https://github.com/MichaelArestad/Press-This/issues/45#issuecomment-74323284 and https://github.com/MichaelArestad/Press-This/issues/45#issuecomment-74324094)

**Comment by MichaelArestad** [comment]
@marcelomazza @folletto @melchoyce 

Maybe something like this? (with underlines on searched characters)

![image](https://cloud.githubusercontent.com/assets/1123119/6195714/d21976b2-b38d-11e4-8ad4-707ceb1f9f06.png)

**Comment by marcelomazza** [comment]
:+1: I just noticed the magnifier icon instead the button "Search"... I'm not sure if the button is clearer.

**Comment by MichaelArestad** [comment]
@marcelomazza 
> I just noticed the magnifier icon instead the button "Search"... I'm not sure if the button is clearer.

I realized there was no need for a button at all since it could search as you type (like on suggested tags). The icon is more of just a label.

**Comment by azaozz** [comment]
Wow, this looks so nice! I'm starting to think we should replace the Categories postbox too :)

Another idea for "flat" listing of cats may be: `parent cat / child cat / grandchild cat`, i.e. separated by slashes. This is how the URL looks (usually) for the categories archive, and most places that use many nested cats do it to get these URL.

**Comment by MichaelArestad** [comment]
Quick stab at @melchoyce 's idea. If the search produces no results, flip into the add category mode.

![image](https://cloud.githubusercontent.com/assets/1123119/6195881/c6598fea-b38f-11e4-8533-4d8ba9d5b159.png)

**Comment by marcelomazza** [comment]
I like the idea of adding on the fly. I would talk with @stephdau to check it, because of the additional logic. But definitely a good idea.

**Comment by stephdau** [comment]
> I would talk with @stephdau to check it

I'd like @azaozz's opinion and advice as well, from an implementation standpoint. With both search and add, for that matter.

**Comment by stephdau** [comment]
Where are we at with the design? Running out of time to implement one. :)

**Comment by MichaelArestad** [comment]
@stephdau The html of the designs above are all pretty much the same. I would prefer we go with the options that show the parent categories as normal selectable categories in the search results like this:

![image](https://cloud.githubusercontent.com/assets/1123119/6216956/08f8aadc-b5cc-11e4-96c5-b33e268c4b2a.png)

What questions do you need answered before you continue?

**Comment by marcelomazza** [comment]
I'm leaving this here as a reminder:

For the :focus and the "selected" style on the categories list, we'd need a change in the HTML. Right know the structure looks like `<label><input>Category</label>`. The input is a checkbox that we hide, and we can't style the `label` because it's not possible to style a parent based on the state of it's child. The ideal structure would be `<input><label>Category</label>`.

On the meantime, we probably can style the `label` toggling a class with JS, based on the state of the input.

**Comment by stephdau** [comment]
Yes, please handle everything on our end, for now, because the category list uses core's `wp_terms_checklist()`, which is highly desirable, over producing our own markup. Once/if we merge into core, we can maybe update the latter function appropriately instead, if still needed and beneficial to all.

Currently, the extra styling for those make it that if one clicks a category entry t check it, it look like the parent category is selected instead: https://cloudup.com/c_NeA1LsOCR

@iseulde also noticed that the category list's scrolling and styling are busted with very long lists (using the [themes unit test data](https://wpcom-themes.svn.automattic.com/demo/theme-unit-test-data.xml)), which I think @MichaelArestad is aware of.

**Comment by stephdau** [comment]
@iseulde has the search on its way, but I've made a call on not implementing the underlining of the matched string feature, in the results, because it implies that the whole lost needs to be redrawn at every character input, jus for that. First, it makes everything way more complex, but I'm also really concerned that the computations involved in this will ultimately lead to a very flickering UX/UI, especially so on slower devices like mobile ones. When coupling both of those concerns, I think it's better if we keep that for a later version, to be experimented with, not for the initial merge.

**Comment by MichaelArestad** [comment]
@stephdau Sounds good. I can live without the underline if it means improved performance.

**Comment by MichaelArestad** [comment]
I think I can go ahead and call this closed. Huzzah!
